src/liquidity_math.rs

```diff
@@ -1,4 +1,115 @@
-use crate::error::UniswapV3MathError;
+use crate::{
+    error::UniswapV3MathError,
+    full_math::mul_div,
+    sqrt_price_math::{FIXED_POINT_96_RESOLUTION, Q96}
+};
+use alloy::primitives::U256;
+
+// Downcasts U256 to u128
+fn to_u128(x: U256) -> Result<u128, UniswapV3MathError> {
+    x.to_string().parse().map_err(|_| UniswapV3MathError::SafeCastToU128Overflow)
+}
+
+// Computes the amount of liquidity received for a given amount of token0 and price range. Returns (uint128 liquidity)
+pub fn get_liquidity_for_amount0(
+    mut sqrt_pa: U256,
+    mut sqrt_pb: U256,
+    amount0: U256,
+) -> Result<u128, UniswapV3MathError> {
+    if sqrt_pa > sqrt_pb { std::mem::swap(&mut sqrt_pa, &mut sqrt_pb) };
+    let intermediate = mul_div(sqrt_pa, sqrt_pb, Q96)?;
+    let liquidity = mul_div(amount0, intermediate, sqrt_pb - sqrt_pa)?;
+
+    to_u128(liquidity)
+}
+
+// Computes the amount of liquidity received for a given amount of token1 and price range. Returns (uint128 liquidity)
+pub fn get_liquidity_for_amount1(
+    mut sqrt_pa: U256,
+    mut sqrt_pb: U256,
+    amount1: U256,
+) -> Result<u128, UniswapV3MathError> {
+    if sqrt_pa > sqrt_pb { std::mem::swap(&mut sqrt_pa, &mut sqrt_pb) };
+    let liquidity = mul_div(amount1, Q96, sqrt_pb - sqrt_pa)?;
+
+    to_u128(liquidity)
+}
```

Comment:
Lets update to avoid `std::mem::swap` and simplify to something like the following. Can you update this in all cases?


```suggestion
pub fn get_liquidity_for_amount1(
    sqrt_pa: U256,
    sqrt_pb: U256,
    amount1: U256,
) -> u128 {
    let denom = if sqrt_pa > sqrt_pb { sqrt_pb - sqrt_pa} else {sqrt_pa -  sqrt_pb};
    let liquidity = mul_div(amount1, Q96, denom)?;

    liquidity::to<u128>()
}
```

Reply:
sure, will do!